aoc2018/src/main/scala/day09.scala

```diff
@@ -0,0 +1,225 @@
+import scala.io.*
+import math.*
+import scala.collection.mutable
+import scala.collection.mutable.Stack
+
+/**
+ *
+ * PART 1:
+ *
+ * This one was pretty easy. I implemented a rotate function for Lists, as well as takeHead and appendHead for taking
+ * the head of a list or appending a value to the head of a list.
+ *
+ * The simulation function just followed the rules as presented in the problem description:
+ *
+ *  0. exit condition --> marble is lastMarble
+ *  1. add new marble
+ *  2. if new marble is divisible by 23 (mod 23) then
+ *    a. rotate marble circle by -7, take marble and compute score as takenMarble + current marble.
+ *  3. else
+ *    a. rotate marble circly by +2 and append new marble to the head of the circle. computed score is 0
+ *  4. update scoreboard with the player being nPlayers % marble with the computed score
+ *  5. recurse
+ *
+ *
+ *  Total computing time was around 12 seconds
+ *
+ *  PART 2:
+ *
+ *  Okay happy days were suddenly over because my implementation to part 1 was not performant. Hence the large amount
+ *  of code in the graveyard at the bottom of this file. I tried working with streams (LazyList in Scala 3) but
+ *  couldn't make it work. I could make a stream of marble locations, but not the actual marble circe because it
+ *  required a state that I couldn't figure out (using LazyList.unfold).
+ *
+ *  So then I just replaced all the Ints with Longs, and the List with Vector. For the latter I updated the implemented
+ *  functions rotate, takeHead and appendHead to work with Vectors. This made the code performant.
+ *
+ *  Lesson learned, use performant data structures when they're available.
+ *
+ */
+
+
+object day09 extends App:
+
+  private val day: String =
+    this.getClass.getName.drop(3).init
+
+  private val start1: Long =
+    System.currentTimeMillis
+
+  private val input: (Int, Int) =
+    Source
+      .fromResource(s"day$day.txt")
+      .getLines
+      .toList
+      .map{
+        case s"${nPlayers} players; last marble is worth ${lastMarblePoints} points" => (nPlayers.toInt, lastMarblePoints.toInt)
+        case _ => sys.error("boom")
+      }
+      .head
+
+  object Vectorr {
+    def takeHeadVector[A](ls: Vector[A]): (A, Vector[A]) =
+      if ls.isEmpty then sys.error("Empty Vector doesn't have head")
+      else
+        val tmp = ls.splitAt(1)
+        (tmp._1.head, tmp._2)
+
+    def appendHeadVector[A](as: Vector[A], r: A): Vector[A] =
+      as.+:(r)
+
+    def rotateVector[A](N: Int, s: Vector[A]): Vector[A] =
+      if s.isEmpty then s
+      else
+        val nbound = N % s.length // skipping the full rotation rounds
+        if nbound < 0 then rotateVector(nbound + s.length, s)
+        else s.drop(nbound) ++ s.take(nbound)
+  }
+
+  def simulateMarbleGame(lastMarble: Int, nPlayers: Int, scoreBoard: Map[Int, Long] = Map(),
```

Comment:
minor style - pref `Map.empty` over `Map()` - note, this counts for all collection types..

( except `SortedMap` as it has no `empty[K,V]` defined 🙄 )